File: internal/protocol/protocol.go
// Package protocol defines the JSON wire messages exchanged over the
// lobby/signaling WebSocket, plus strict validation for everything a
// client can send. The server never trusts a client-supplied field
// without it passing through this package first.
package protocol

import (
	"encoding/json"
	"errors"
	"fmt"
	"time"
)

// Message type strings, client -> server.
const (
	TypeJoin      = "join"
	TypeClaimSlot = "claim-slot"
	TypeSignal    = "signal"
	TypeLeave     = "leave"
)

// Message type strings, server -> client.
const (
	TypeJoined         = "joined"
	TypePlayerJoined   = "player-joined"
	TypePlayerLeft     = "player-left"
	TypePlayerRejoined = "player-rejoined"
	TypePlayerReplaced = "player-replaced"
	TypeRoomStarted    = "room-started"
	TypeError          = "error"
)

// Error codes carried in Error.Code. Stable API: clients switch on these.
const (
	ErrCodeInvalidMessage    = "invalid-message"
	ErrCodeInvalidCode       = "invalid-code"
	ErrCodeInvalidCreate     = "invalid-create"
	ErrCodeUnknownApp        = "unknown-app"
	ErrCodeAppMismatch       = "app-mismatch"
	ErrCodeOriginForbidden   = "origin-forbidden"
	ErrCodeRoomNotFound      = "room-not-found"
	ErrCodeRoomFull          = "room-full"
	ErrCodeRoomStarted       = "room-started"
	ErrCodeTooManyRooms      = "too-many-rooms"
	ErrCodeAlreadyJoined     = "already-joined"
	ErrCodeNotJoined         = "not-joined"
	ErrCodeInvalidTarget     = "invalid-target"
	ErrCodeTargetUnavailable = "target-unavailable"
	ErrCodeClaimNotAllowed   = "claim-not-allowed"
	ErrCodeSlotNotClaimable  = "slot-not-claimable"
	ErrCodeReplaced          = "replaced"
	ErrCodeSuperseded        = "session-superseded"
	ErrCodeRoomExpired       = "room-expired"
	ErrCodeSlowConsumer      = "slow-consumer"
	ErrCodeInternal          = "internal"
)

// Player-left reasons.
const (
	LeftReasonExplicit     = "explicit-leave"
	LeftReasonDisconnected = "disconnected"
)

// ReconnectPolicy controls how a slot may be re-acquired.
type ReconnectPolicy string

const (
	PolicyTokenOnly                ReconnectPolicy = "token-only"
	PolicyTokenOrClaimAfterTimeout ReconnectPolicy = "token-or-claim-after-timeout"
	PolicyClaimAfterTimeout        ReconnectPolicy = "claim-after-timeout"
	// PolicyHostApproval is reserved in the spec but not implemented; the
	// server rejects room creation that requests it.
	PolicyHostApproval ReconnectPolicy = "host-approval"
)

// Valid reports whether p is a policy this server implements.
func (p ReconnectPolicy) Valid() bool {
	switch p {
	case PolicyTokenOnly, PolicyTokenOrClaimAfterTimeout, PolicyClaimAfterTimeout:
		return true
	}
	return false
}

// AllowsClaim reports whether the policy permits tokenless slot claiming.
func (p ReconnectPolicy) AllowsClaim() bool {
	return p == PolicyTokenOrClaimAfterTimeout || p == PolicyClaimAfterTimeout
}

// DefaultReconnectPolicy is used when room creation omits the policy.
const DefaultReconnectPolicy = PolicyTokenOrClaimAfterTimeout

// Hard bounds on room codes (spec: length 4-64, safe charset only).
const (
	MinCodeLen = 4
	MaxCodeLen = 64
)

// Signal payload kinds a client may relay.
const (
	SignalKindOffer  = "offer"
	SignalKindAnswer = "answer"
	SignalKindICE    = "ice"
)

// ProtoError is a protocol-level failure with a stable machine-readable
// code. It is returned by validation and lobby operations and mapped
// directly onto an Error message.
type ProtoError struct {
	Code    string
	Message string
}

func (e *ProtoError) Error() string { return e.Code + ": " + e.Message }

// Errf builds a *ProtoError with a formatted message.
func Errf(code, format string, args ...any) *ProtoError {
	return &ProtoError{Code: code, Message: fmt.Sprintf(format, args...)}
}

// ClientMessage is the envelope for every client -> server message.
// Unknown JSON fields are ignored for forward compatibility; known
// fields are strictly validated per message type.
type ClientMessage struct {
	Type        string          `json:"type"`
	AppID       string          `json:"appId,omitempty"`
	Code        string          `json:"code,omitempty"`
	ResumeToken string          `json:"resumeToken,omitempty"`
	Create      *CreateOptions  `json:"create,omitempty"`
	PlayerID    *int            `json:"playerId,omitempty"`
	To          *int            `json:"to,omitempty"`
	Payload     json.RawMessage `json:"payload,omitempty"`
}

// CreateOptions are the client-supplied room creation options. Pointer
// fields distinguish "absent" from "false"/zero so defaults can apply.
type CreateOptions struct {
	MaxPlayers       int              `json:"maxPlayers"`
	WaitUntilFull    *bool            `json:"waitUntilFull,omitempty"`
	AllowLateJoin    *bool            `json:"allowLateJoin,omitempty"`
	AllowReconnect   *bool            `json:"allowReconnect,omitempty"`
	AllowReplacement *bool            `json:"allowReplacement,omitempty"`
	ReconnectPolicy  *ReconnectPolicy `json:"reconnectPolicy,omitempty"`
	ClaimAfterMs     *int64           `json:"claimAfterMs,omitempty"`
}

// RoomOptions is the fully-resolved, validated form of CreateOptions.
type RoomOptions struct {
	MaxPlayers       int
	WaitUntilFull    bool
	AllowLateJoin    bool
	AllowReconnect   bool
	AllowReplacement bool
	ReconnectPolicy  ReconnectPolicy
	ClaimAfter       time.Duration
}

// ResolveCreate validates opts against hard limits and fills defaults.
// hardMaxPlayers caps maxPlayers; defaultClaimAfter applies when the
// client omits claimAfterMs.
//
// Defaults: waitUntilFull=false, allowLateJoin=true, allowReconnect=true,
// allowReplacement=true, policy=token-or-claim-after-timeout. Note
// allowLateJoin defaults true so that a default room (which starts
// immediately because waitUntilFull=false) remains joinable.
func ResolveCreate(opts *CreateOptions, hardMaxPlayers int, defaultClaimAfter time.Duration) (RoomOptions, error) {
	if opts == nil {
		return RoomOptions{}, Errf(ErrCodeInvalidCreate, "create options required")
	}
	if opts.MaxPlayers < 1 {
		return RoomOptions{}, Errf(ErrCodeInvalidCreate, "maxPlayers must be >= 1")
	}
	if opts.MaxPlayers > hardMaxPlayers {
		return RoomOptions{}, Errf(ErrCodeInvalidCreate, "maxPlayers %d exceeds limit %d", opts.MaxPlayers, hardMaxPlayers)
	}
	r := RoomOptions{
		MaxPlayers:       opts.MaxPlayers,
		WaitUntilFull:    boolOr(opts.WaitUntilFull, false),
		AllowLateJoin:    boolOr(opts.AllowLateJoin, true),
		AllowReconnect:   boolOr(opts.AllowReconnect, true),
		AllowReplacement: boolOr(opts.AllowReplacement, true),
		ReconnectPolicy:  DefaultReconnectPolicy,
		ClaimAfter:       defaultClaimAfter,
	}
	if opts.ReconnectPolicy != nil {
		if !opts.ReconnectPolicy.Valid() {
			return RoomOptions{}, Errf(ErrCodeInvalidCreate, "unsupported reconnectPolicy %q", *opts.ReconnectPolicy)
		}
		r.ReconnectPolicy = *opts.ReconnectPolicy
	}
	if opts.ClaimAfterMs != nil {
		ms := *opts.ClaimAfterMs
		if ms < 0 {
			return RoomOptions{}, Errf(ErrCodeInvalidCreate, "claimAfterMs must be >= 0")
		}
		if ms > int64(24*time.Hour/time.Millisecond) {
			return RoomOptions{}, Errf(ErrCodeInvalidCreate, "claimAfterMs exceeds 24h")
		}
		r.ClaimAfter = time.Duration(ms) * time.Millisecond
	}
	return r, nil
}

func boolOr(p *bool, def bool) bool {
	if p != nil {
		return *p
	}
	return def
}

// ValidateCode enforces the room code rules: 4-64 chars from
// [A-Za-z0-9_-]. Codes are case-sensitive and matched exactly.
func ValidateCode(code string) error {
	if len(code) < MinCodeLen || len(code) > MaxCodeLen {
		return Errf(ErrCodeInvalidCode, "room code must be %d-%d characters", MinCodeLen, MaxCodeLen)
	}
	for i := 0; i < len(code); i++ {
		c := code[i]
		switch {
		case c >= 'A' && c <= 'Z', c >= 'a' && c <= 'z', c >= '0' && c <= '9', c == '_', c == '-':
		default:
			return Errf(ErrCodeInvalidCode, "room code contains invalid character %q", string(c))
		}
	}
	return nil
}

// MaxResumeTokenLen bounds the resumeToken field (real tokens are 43
// chars of base64url; anything much longer is garbage).
const MaxResumeTokenLen = 128

// ParseClientMessage decodes and structurally validates a client
// message. It does not check room state; the lobby does that.
func ParseClientMessage(data []byte) (*ClientMessage, error) {
	var m ClientMessage
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, Errf(ErrCodeInvalidMessage, "malformed JSON: %v", err)
	}
	switch m.Type {
	case TypeJoin:
		if err := ValidateCode(m.Code); err != nil {
			return nil, err
		}
		if len(m.ResumeToken) > MaxResumeTokenLen {
			return nil, Errf(ErrCodeInvalidMessage, "resumeToken too long")
		}
	case TypeClaimSlot:
		if err := ValidateCode(m.Code); err != nil {
			return nil, err
		}
		if m.PlayerID == nil {
			return nil, Errf(ErrCodeInvalidMessage, "claim-slot requires playerId")
		}
	case TypeSignal:
		if m.To == nil {
			return nil, Errf(ErrCodeInvalidMessage, "signal requires 'to'")
		}
		if err := validateSignalPayload(m.Payload); err != nil {
			return nil, err
		}
	case TypeLeave:
		// no fields
	case "":
		return nil, Errf(ErrCodeInvalidMessage, "missing message type")
	default:
		return nil, Errf(ErrCodeInvalidMessage, "unknown message type %q", m.Type)
	}
	return &m, nil
}

func validateSignalPayload(raw json.RawMessage) error {
	if len(raw) == 0 {
		return Errf(ErrCodeInvalidMessage, "signal requires payload")
	}
	var peek struct {
		Kind string `json:"kind"`
	}
	if err := json.Unmarshal(raw, &peek); err != nil {
		return Errf(ErrCodeInvalidMessage, "signal payload must be a JSON object: %v", err)
	}
	switch peek.Kind {
	case SignalKindOffer, SignalKindAnswer, SignalKindICE:
		return nil
	default:
		return Errf(ErrCodeInvalidMessage, "signal payload kind must be offer, answer, or ice")
	}
}

// PlayerInfo is the public snapshot of one slot, sent in joined and
// player-joined messages. LastSeenMsAgo is only meaningful when
// Occupied is true.
type PlayerInfo struct {
	ID            int   `json:"id"`
	Occupied      bool  `json:"occupied"`
	Connected     bool  `json:"connected"`
	LastSeenMsAgo int64 `json:"lastSeenMsAgo"`
}

// ICEServer mirrors the WebRTC RTCIceServer dictionary.
type ICEServer struct {
	URLs       []string `json:"urls"`
	Username   string   `json:"username,omitempty"`
	Credential string   `json:"credential,omitempty"`
}

// Joined is the successful response to join or claim-slot.
type Joined struct {
	Type        string       `json:"type"`
	Code        string       `json:"code"`
	SelfID      int          `json:"selfId"`
	MaxPlayers  int          `json:"maxPlayers"`
	Started     bool         `json:"started"`
	ResumeToken string       `json:"resumeToken"`
	Players     []PlayerInfo `json:"players"`
	ICEServers  []ICEServer  `json:"iceServers,omitempty"`
}

// PlayerJoined announces a fresh occupant of a previously empty slot.
type PlayerJoined struct {
	Type     string       `json:"type"`
	PlayerID int          `json:"playerId"`
	Players  []PlayerInfo `json:"players"`
}

// PlayerLeft announces either an explicit leave (slot freed) or a
// transport disconnect (slot retained); Reason distinguishes.
type PlayerLeft struct {
	Type     string `json:"type"`
	PlayerID int    `json:"playerId"`
	Reason   string `json:"reason"`
}

// PlayerRejoined announces a token-based resume of an occupied slot.
type PlayerRejoined struct {
	Type           string `json:"type"`
	PlayerID       int    `json:"playerId"`
	WasReplacement bool   `json:"wasReplacement"`
}

// PlayerReplaced announces a tokenless claim of a silent occupied slot.
type PlayerReplaced struct {
	Type     string `json:"type"`
	PlayerID int    `json:"playerId"`
}

// RoomStarted announces the started transition.
type RoomStarted struct {
	Type string `json:"type"`
}

// SignalIn is a relayed signaling payload, server -> client.
type SignalIn struct {
	Type    string          `json:"type"`
	From    int             `json:"from"`
	Payload json.RawMessage `json:"payload"`
}

// Error is the server -> client error message.
type Error struct {
	Type    string `json:"type"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

// ErrorMessage builds an Error from any error, mapping *ProtoError
// codes through and everything else to "internal".
func ErrorMessage(err error) Error {
	var pe *ProtoError
	if errors.As(err, &pe) {
		return Error{Type: TypeError, Code: pe.Code, Message: pe.Message}
	}
	return Error{Type: TypeError, Code: ErrCodeInternal, Message: "internal error"}
}

File: clients/rust/src/core/protocol.rs
//! Signaling wire types (JSON over the WebSocket). Field and tag
//! names must match the server / TS client exactly.

use serde::{Deserialize, Deserializer, Serialize};

use super::options::CreateOptions;
use super::roster::PlayerId;

/// Server error codes after which the WebSocket will not come back.
pub fn is_fatal_code(code: &str) -> bool {
    matches!(
        code,
        "replaced" | "session-superseded" | "room-expired" | "slow-consumer"
    )
}

/// Fatal codes that also mean our peers are gone / we left the room.
pub fn is_game_over_code(code: &str) -> bool {
    matches!(code, "replaced" | "session-superseded" | "room-expired")
}

// ---------------------------------------------------------------------------
// Client -> server
// ---------------------------------------------------------------------------

#[derive(Debug, Serialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum ClientMessage<'a> {
    Join {
        code: &'a str,
        #[serde(rename = "appId", skip_serializing_if = "Option::is_none")]
        app_id: Option<&'a str>,
        #[serde(rename = "resumeToken", skip_serializing_if = "Option::is_none")]
        resume_token: Option<&'a str>,
        #[serde(skip_serializing_if = "Option::is_none")]
        create: Option<&'a CreateOptions>,
    },
    ClaimSlot {
        code: &'a str,
        #[serde(rename = "playerId")]
        player_id: PlayerId,
        #[serde(rename = "appId", skip_serializing_if = "Option::is_none")]
        app_id: Option<&'a str>,
    },
    Signal {
        to: PlayerId,
        payload: &'a SignalPayload,
    },
    Leave,
}

pub fn encode_client_message(msg: &ClientMessage<'_>) -> String {
    // Our own types cannot fail to serialize.
    serde_json::to_string(msg).expect("client message serialization")
}

// ---------------------------------------------------------------------------
// Server -> client
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Deserialize)]
pub struct WirePlayer {
    pub id: PlayerId,
    #[serde(default)]
    pub occupied: bool,
    #[serde(default)]
    pub connected: bool,
    // lastSeenMsAgo is ignored (unknown fields are skipped by serde).
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JoinedMsg {
    pub code: String,
    pub self_id: PlayerId,
    pub max_players: u16,
    #[serde(default)]
    pub started: bool,
    pub resume_token: String,
    pub players: Vec<WirePlayer>,
    #[serde(default)]
    pub ice_servers: Vec<IceServer>,
}

/// WebRTC signal relayed through the lobby. The ICE candidate is kept
/// as raw JSON (RTCIceCandidateInit); clients pass it through without
/// interpreting it. `candidate` null/absent = end-of-candidates.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum SignalPayload {
    Offer {
        sdp: String,
    },
    Answer {
        sdp: String,
    },
    Ice {
        #[serde(skip_serializing_if = "Option::is_none")]
        candidate: Option<serde_json::Value>,
    },
    /// Unknown kinds are ignored for forward compatibility.
    #[serde(other)]
    Unknown,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum ServerMessage {
    Joined(JoinedMsg),
    #[serde(rename_all = "camelCase")]
    PlayerJoined {
        player_id: PlayerId,
        players: Vec<WirePlayer>,
    },
    #[serde(rename_all = "camelCase")]
    PlayerLeft {
        player_id: PlayerId,
        reason: String,
    },
    #[serde(rename_all = "camelCase")]
    PlayerRejoined {
        player_id: PlayerId,
        was_replacement: bool,
    },
    #[serde(rename_all = "camelCase")]
    PlayerReplaced {
        player_id: PlayerId,
    },
    RoomStarted,
    Signal {
        from: PlayerId,
        payload: SignalPayload,
    },
    Error {
        code: String,
        message: String,
    },
    /// Unknown message types are ignored for forward compatibility.
    #[serde(other)]
    Unknown,
}

// ---------------------------------------------------------------------------
// ICE servers
// ---------------------------------------------------------------------------

/// One RTCIceServer entry as issued by the server (or supplied by the
/// user). `urls` accepts a single string or an array on the wire.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct IceServer {
    #[serde(deserialize_with = "string_or_seq", default)]
    pub urls: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub credential: Option<String>,
}

fn string_or_seq<'de, D: Deserializer<'de>>(d: D) -> Result<Vec<String>, D::Error> {
    #[derive(Deserialize)]
    #[serde(untagged)]
    enum OneOrMany {
        One(String),
        Many(Vec<String>),
    }
    Ok(match OneOrMany::deserialize(d)? {
        OneOrMany::One(s) => vec![s],
        OneOrMany::Many(v) => v,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn client_messages_encode_like_ts() {
        let create = CreateOptions::new(4);
        let msg = ClientMessage::Join {
            code: "ROOM",
            app_id: None,
            resume_token: Some("tok"),
            create: Some(&create),
        };
        let v: serde_json::Value =
            serde_json::from_str(&encode_client_message(&msg)).unwrap();
        assert_eq!(v["type"], "join");
        assert_eq!(v["code"], "ROOM");
        assert_eq!(v["resumeToken"], "tok");
        assert_eq!(v["create"]["maxPlayers"], 4);
        assert_eq!(v["create"]["waitUntilFull"], false);
        assert_eq!(v["create"]["allowLateJoin"], true);
        assert_eq!(v["create"]["allowReconnect"], true);
        assert_eq!(v["create"]["allowReplacement"], true);
        assert!(v["create"].get("reconnectPolicy").is_none());
        assert!(v["create"].get("claimAfterMs").is_none());
        assert!(v.get("appId").is_none());

        let msg = ClientMessage::ClaimSlot {
            code: "ROOM",
            player_id: 2,
            app_id: Some("app"),
        };
        let v: serde_json::Value =
            serde_json::from_str(&encode_client_message(&msg)).unwrap();
        assert_eq!(v["type"], "claim-slot");
        assert_eq!(v["playerId"], 2);
        assert_eq!(v["appId"], "app");

        let payload = SignalPayload::Offer { sdp: "SDP".into() };
        let msg = ClientMessage::Signal { to: 1, payload: &payload };
        let v: serde_json::Value =
            serde_json::from_str(&encode_client_message(&msg)).unwrap();
        assert_eq!(v["type"], "signal");
        assert_eq!(v["to"], 1);
        assert_eq!(v["payload"]["kind"], "offer");
        assert_eq!(v["payload"]["sdp"], "SDP");

        assert_eq!(
            encode_client_message(&ClientMessage::Leave),
            r#"{"type":"leave"}"#
        );
    }

    #[test]
    fn reconnect_policy_strings() {
        let mut create = CreateOptions::new(2);
        create.reconnect_policy = Some(crate::core::options::ReconnectPolicy::TokenOrClaimAfterTimeout);
        create.claim_after_ms = Some(40_000);
        let v = serde_json::to_value(&create).unwrap();
        assert_eq!(v["reconnectPolicy"], "token-or-claim-after-timeout");
        assert_eq!(v["claimAfterMs"], 40_000);
    }

    #[test]
    fn parses_joined() {
        let raw = r#"{"type":"joined","code":"ROOM","selfId":0,"maxPlayers":2,
            "started":false,"resumeToken":"tok",
            "players":[{"id":0,"occupied":true,"connected":true,"lastSeenMsAgo":0},
                       {"id":1,"occupied":false,"connected":false}],
            "iceServers":[{"urls":"stun:stun.example.com"},
                          {"urls":["turn:t.example.com"],"username":"u","credential":"c"}]}"#;
        let msg: ServerMessage = serde_json::from_str(raw).unwrap();
        let ServerMessage::Joined(j) = msg else {
            panic!("expected joined")
        };
        assert_eq!(j.self_id, 0);
        assert_eq!(j.max_players, 2);
        assert_eq!(j.resume_token, "tok");
        assert_eq!(j.players.len(), 2);
        assert!(j.players[0].occupied && j.players[0].connected);
        assert_eq!(j.ice_servers[0].urls, vec!["stun:stun.example.com"]);
        assert_eq!(j.ice_servers[1].urls, vec!["turn:t.example.com"]);
        assert_eq!(j.ice_servers[1].username.as_deref(), Some("u"));
    }

    #[test]
    fn parses_joined_without_ice_servers() {
        let raw = r#"{"type":"joined","code":"ROOM","selfId":1,"maxPlayers":2,
            "started":true,"resumeToken":"t","players":[]}"#;
        let ServerMessage::Joined(j) = serde_json::from_str(raw).unwrap() else {
            panic!()
        };
        assert!(j.started);
        assert!(j.ice_servers.is_empty());
    }

    #[test]
    fn parses_server_messages() {
        let m: ServerMessage = serde_json::from_str(
            r#"{"type":"player-left","playerId":1,"reason":"explicit-leave"}"#,
        )
        .unwrap();
        assert!(matches!(
            m,
            ServerMessage::PlayerLeft { player_id: 1, ref reason } if reason == "explicit-leave"
        ));

        let m: ServerMessage = serde_json::from_str(
            r#"{"type":"player-rejoined","playerId":0,"wasReplacement":true}"#,
        )
        .unwrap();
        assert!(matches!(
            m,
            ServerMessage::PlayerRejoined { player_id: 0, was_replacement: true }
        ));

        let m: ServerMessage = serde_json::from_str(r#"{"type":"room-started"}"#).unwrap();
        assert!(matches!(m, ServerMessage::RoomStarted));

        let m: ServerMessage = serde_json::from_str(
            r#"{"type":"signal","from":1,"payload":{"kind":"ice","candidate":{"candidate":"x","sdpMid":"0"}}}"#,
        )
        .unwrap();
        let ServerMessage::Signal { from: 1, payload: SignalPayload::Ice { candidate: Some(c) } } = m
        else {
            panic!("expected ice signal")
        };
        assert_eq!(c["sdpMid"], "0");

        // null candidate -> None (end-of-candidates)
        let m: ServerMessage = serde_json::from_str(
            r#"{"type":"signal","from":1,"payload":{"kind":"ice","candidate":null}}"#,
        )
        .unwrap();
        assert!(matches!(
            m,
            ServerMessage::Signal { payload: SignalPayload::Ice { candidate: None }, .. }
        ));
    }

    #[test]
    fn unknown_types_are_forward_compatible() {
        let m: ServerMessage =
            serde_json::from_str(r#"{"type":"totally-new","x":1}"#).unwrap();
        assert!(matches!(m, ServerMessage::Unknown));
        let m: ServerMessage = serde_json::from_str(
            r#"{"type":"signal","from":0,"payload":{"kind":"future-thing","data":1}}"#,
        )
        .unwrap();
        assert!(matches!(
            m,
            ServerMessage::Signal { payload: SignalPayload::Unknown, .. }
        ));
    }

    #[test]
    fn fatal_code_sets() {
        for c in ["replaced", "session-superseded", "room-expired", "slow-consumer"] {
            assert!(is_fatal_code(c));
        }
        assert!(!is_fatal_code("room-full"));
        assert!(is_game_over_code("room-expired"));
        assert!(!is_game_over_code("slow-consumer"));
    }
}
